components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/obsrvability/handler/metrics/publisher/MetricHandler.java

```diff
@@ -0,0 +1,480 @@
+
+    @Override
+    public boolean handleInit() {
+        getMetricReporter();
+        metricReporterInstance.initMetrics();
+
+        String host = System.getProperty(MicroIntegratorBaseConstants.LOCAL_IP_ADDRESS);
+        String port = System.getProperty(MetricConstants.HTTP_PORT);
+        String javaVersion = System.getProperty(MetricConstants.JAVA_VERSION);
+        String javaHome = System.getProperty(MetricConstants.JAVA_HOME);
+
+        metricReporterInstance.serverUp(host, port, javaVersion, javaHome);
+        return true;
+    }
+
+    /**
+     * Load the MetricReporter class from the deployment.toml file if a user has defined a MetricReporter.
+     * Use default PrometheusReporter if the user hasn't defined a MetricReporter or an error occurs
+     * during custom MetricReporter class invocation.
+     */
+    private void getMetricReporter() {
+        Map<String, Object> configs = ConfigParser.getParsedConfigs();
+        Object metricReporterClass = configs.get(MetricConstants.PROMETHEUS_HANDLER + "." + METRIC_REPORTER);
+        log.info("Loading the class " + metricReporterClass);
+
+        if (metricReporterClass != null) {
+            try {
+                loadedMetricClass = Class.forName(metricReporterClass.toString());
+                metricReporterInstance = (MetricReporter) loadedMetricClass.newInstance();
+            } catch (Exception e) {
+                log.error("Error in loading the class" + metricReporterClass.toString() +
+                        "Hence loading the default PrometheusReporter class ");
+                loadPrometheusReporter();
+            }
+        } else {
+            loadPrometheusReporter();
+        }
+    }
+
+    /**
+     * Load the PrometheusReporter class by default.
+     */
+    private void loadPrometheusReporter() {
+        try {
+            loadedMetricClass = Class.forName("org.wso2.micro.integrator.obsrvability.handler.metrics.publisher" +
+                    ".prometheus.reporter.PrometheusReporter");
+            metricReporterInstance = (MetricReporter) loadedMetricClass.newInstance();
+        } catch (Exception e) {
+            handleException(e,"Error in loading the PrometheusReporter Class");
+        }
+    }
+
+    @Override
+    public boolean handleRequestInFlow(MessageContext synCtx) {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx)
+                .getAxis2MessageContext();
+
+        int portOffset;
+        int internalHttpApiPort = 0;
+        int serviceInvokePort = 0;
+
+        // Get the Internal HTTP Inbound Endpoint port
+        if ((null != System.getProperty(MetricConstants.PORT_OFFSET))) {
+            portOffset = Integer.parseInt(System.getProperty((MetricConstants.PORT_OFFSET)));
+            internalHttpApiPort = Integer.parseInt(synCtx.getEnvironment().getSynapseConfiguration().
+                    getProperty((MetricConstants.INTERNAL_HTTP_API_PORT)));
```

Comment:
these are constants and we shouldn't be doing this for all requests.

Reply:
Changed the reading of System properties to be done only once.